contracts/0.8.25/vaults/Sideloading.sol

```diff
@@ -0,0 +1,324 @@
+     *
+     * NB: Sideloading conservatively assumes that the sideloader will not be able to
+     *     obtain more ETH than the amount of stETH minted. This is why the initial valuation
+     *     must cover the minimal reserve after sideloading.
+     *     For example, if the vault reserve ratio is 10% and the vault owner wants to mint 900 stETH,
+     *     the initial valuation of the vault must be at least 100 ETH (10% of 1000 ETH).
+     *
+     * @param _vault The address of the vault.
+     * @param _sideloader The address of the sideloader.
+     * @param _amountOfShares The amount of shares to mint for sideloading.
+     * @param _data The data to pass to the sideloader.
+     * @return True if the sideload was successful.
+     */
+    function sideload(
+        address _vault,
+        address _sideloader,
+        uint256 _amountOfShares,
+        bytes calldata _data
+    ) external whenResumed returns (bool) {
+        // * * * input validation * * *
+        if (_vault == address(0)) revert ZeroArgument("_vault");
+        if (_amountOfShares == 0) revert ZeroArgument("_amountOfShares");
+        _vaultAuth(_vault, "sideload");
+
+        SideloadingStorage storage $ = _getSideloadingStorage();
+        if (!$.isSideloaderRegistryIgnored && !$.sideloaderRegistry[_sideloader]) {
+            revert SideloaderNotRegistered(_sideloader);
+        }
+
+        // * * * sufficiency checks * * *
+        VaultSocket storage socket = _connectedSocket(_vault);
+
+        uint256 totalSharesAfterSideload = socket.sharesMinted + _amountOfShares;
+
+        // cannot result in shares exceeding the share limit
+        if (totalSharesAfterSideload > socket.shareLimit) revert ShareLimitExceeded(_vault, socket.shareLimit);
+
+        uint256 totalStETHAfterSideload = STETH.getPooledEthByShares(totalSharesAfterSideload);
+        uint256 minimalReserveAfterSideload = (totalStETHAfterSideload * socket.reserveRatioBP) /
+            (TOTAL_BASIS_POINTS - socket.reserveRatioBP);
+
+        // ensure the valuation before sideloading is sufficient to cover the minimal reserve after sideloading
+        if (IStakingVault(_vault).valuation() < minimalReserveAfterSideload) {
+            revert InsufficientValuationBeforeSideload(
+                _vault,
+                minimalReserveAfterSideload,
+                IStakingVault(_vault).valuation()
+            );
+        }
+
+        uint256 totalEtherLocked = totalStETHAfterSideload + minimalReserveAfterSideload;
+
+        // update the shares minted BEFORE sideloading
+        socket.sharesMinted = uint96(totalSharesAfterSideload);
+
+        // update the locked amount BEFORE sideloading
+        if (totalEtherLocked > IStakingVault(_vault).locked()) {
+            IStakingVault(_vault).lock(totalEtherLocked);
+        }
+
+        // mint the shares to the sideloader
+        STETH.mintExternalShares(_sideloader, _amountOfShares);
```

Comment:
maybe extract (not critical)

Reply:
consider future operator grid and migration

Reply:
will be done separately, once grid and migration is implemented